lib/crypto/sig_verifier_test.go

```diff
@@ -0,0 +1,222 @@
+// Copyright 2021 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package crypto_test
+
+import (
+	"testing"
+
+	"github.com/ChainSafe/gossamer/internal/log"
+	"github.com/ChainSafe/gossamer/lib/crypto"
+	"github.com/ChainSafe/gossamer/lib/crypto/ed25519"
+	"github.com/ChainSafe/gossamer/lib/crypto/secp256k1"
+	"github.com/ChainSafe/gossamer/lib/crypto/sr25519"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestSigVerifierEd25519(t *testing.T) {
+	t.Helper()
+	signs := make([]*crypto.Signing, 2)
+
+	for i := 0; i < 2; i++ {
+		msg := []byte("Hello")
+		key, err := ed25519.GenerateKeypair()
+		require.NoError(t, err)
+
+		sign, err := key.Sign(msg)
+		require.NoError(t, err)
+
+		signs[i] = &crypto.Signing{
+			PubKey:     key.Public().Encode(),
+			Sign:       sign,
+			Msg:        msg,
+			VerifyFunc: ed25519.VerifySignature,
+		}
+	}
+	signVerify := crypto.NewSignatureVerifier(log.New())
+
+	for _, sig := range signs {
+		signVerify.Add(sig)
+	}
+
+	signVerify.Start()
+	ok := signVerify.Finish()
+	require.True(t, ok)
+}
+
+func TestSigVerifierEd25519Fails(t *testing.T) {
+	t.Helper()
+	signs := make([]*crypto.Signing, 2)
+
+	for i := 0; i < 2; i++ {
+		msg := []byte("Hello")
+		key, err := ed25519.GenerateKeypair()
+		require.NoError(t, err)
+
+		sign, err := key.Sign(msg)
+		require.NoError(t, err)
+		if i == 1 {
+			signs[i] = &crypto.Signing{
+				PubKey:     []byte(""),
+				Sign:       sign,
+				Msg:        msg,
+				VerifyFunc: ed25519.VerifySignature,
+			}
+			continue
+		}
+		signs[i] = &crypto.Signing{
+			PubKey:     key.Public().Encode(),
+			Sign:       []byte(""),
+			Msg:        msg,
+			VerifyFunc: ed25519.VerifySignature,
+		}
+	}
+	signVerify := crypto.NewSignatureVerifier(log.New())
+
+	for _, sig := range signs {
+		signVerify.Add(sig)
+	}
+
+	signVerify.Start()
+	ok := signVerify.Finish()
+	require.False(t, ok)
+}
+
+func TestSigVerifierSr25519(t *testing.T) {
+	t.Helper()
+	signs := make([]*crypto.Signing, 2)
+
+	for i := 0; i < 2; i++ {
+		msg := []byte("Hello")
+		key, err := sr25519.GenerateKeypair()
+		require.NoError(t, err)
+
+		sign, err := key.Sign(msg)
+		require.NoError(t, err)
+
+		signs[i] = &crypto.Signing{
+			PubKey:     key.Public().Encode(),
+			Sign:       sign,
+			Msg:        msg,
+			VerifyFunc: sr25519.VerifySignature,
+		}
+	}
+	signVerify := crypto.NewSignatureVerifier(log.New())
+
+	for _, sig := range signs {
+		signVerify.Add(sig)
+	}
+
+	signVerify.Start()
+	ok := signVerify.Finish()
+	require.True(t, ok)
+}
+
+func TestSigVerifierSr25519Fails(t *testing.T) {
```

Comment:
Can you perhaps merge the success and fail tests together? :thinking: It looks like quite a bit of code duplication

Reply:
A had a little trouble trying to use test cases and sub tests in this case.

While doing it I had some problems to infer the keypair type, for example, to dynamically call `type.VerifySignature`, ended up looking more verbose than doing them apart.

Reply:
here use a `map[string]struct{expect err, signaturesToVerify []crypto.SignatureInfo}` then you set manually a success case which has signatures and expect no err and a case error which has wrong signature info data and expects an error. To resolve the problem of keypair generation, I would suggest creating a keypair at the beginning of the function and then use them as you want while constructing the test case map

Reply:
thanks @EclesioMeloJunior for the hint on how to solve it, u are the best!